### ISSUE TITLE
Improve parallel tests concurrency

### ISSUE THREAD

**Issue by tomwidmer** [gradle/gradle#2669]
When running unit tests in parallel using maxParallelForks, tests are allocated to forks statically, using a round-robin algorithm, on startup. This means that there is a chance that all your slow tests will be allocated to 1 fork, and as a result using more than 1 fork will have almost no time saving over running a single fork. 

### Expected Behavior
While there are test classes that haven't yet started, no fork should be idle.

### Current Behavior
You can have a whole series of tests run without any parallelism, if more slow tests happen to be allocated to a particular fork than to others, causing the run to take much longer than it should.

### Context
This has been an issue with long running Selenium tests (> 1 minute each) running against cloud selenium providers who offer a fixed number of simultaneous tests. Test run time is often 50+% longer than it should be, due to it ending up running just 1 test at a time towards the end of the test suite.

### Steps to Reproduce (for bugs)
Create 2 slow test classes (1 minute each) and 2 fast test classes (<1 second). Run them with maxParallelForks = 2. Depending on how the classes are allocated to forks, runtime will be either 1 minute or 2 minutes, a 100% difference.

### Your Environment
Selenium testing using cloud providers (with limited parallelism)
Mac and Linux

**Comment by tomwidmer** [comment]
I can see where the fix needs to be - in MaxNParallelTestClassProcessor. However, I don't know how to wait for an idle TestClassProcessor. Any hints?

**Comment by oehme** [comment]
It wouldn't be that easy to change, because tests are scheduled asynchronously to get the best throughput when you have many short-running ones. The tests are just put in the worker's incoming queue, which it then processes. I.e. the scheduling happens rapidly right at the start and doesn't wait for any response from the workers. This could certainly be improved, but we need to make sure to not make short-running tests inefficient in the process.

A workaround would be to categorize your tests into a fast and slow bucket and have separate tasks for them.

**Comment by dellgreen** [comment]
I'm seeing this as well, some forked jvm's become idle with remaining tests scheduled on busy ones and not making use of free resources available. Pretty cool now that its easy to visualize this now in 4.1 :)

**Comment by tcfurrer** [comment]
Any idea when this issue might get some development action?

I ask because this issue can be a real disappointment for developers who are migrating to Gradle specifically to benefit from its excellent parallelism. 

Especially as both the number of workers and test class runtime increases, the unnecessary increase in wall clock time due to this issue can become embarrassingly bad.  My daily work server has 32 threads, with some individual tests taking hours.  So this issue can result in a real productivity loss for my team.  I guess with such long running tests, I could play around with forkEvery to get some benefit... but that shouldn't be required.

**Comment by 4n00p** [comment]
I'd be interested in helping with development.

**Comment by 4n00p** [comment]
I made a simple quick ‘n’ dirty python script to visualize the test task and Test Executors with wall clock time. It might be useful to some folks here. It wouldve been nice to see which Test Executor belongs to which worker, but I couldnt get that from the debug log statements. 
  
https://github.com/4n00p/gradle_log_test_timeline

It has been useful for visualizing the build I work with.  I made a few projects for demo purposes and put the “screenshots” directory in the proj I linked above.

A few observations:

- There is optimal speedup when the tests are all the same duration. Going from 1 process to 2 maxParallelForks speeds up the task 50%. Note the step ladder shape in screenshots/Figure_3.png and Figure_4.png 

- There is sort of a long tail...towards the end of the task, there are finished workers waiting on other to finish their queue of tests.  Work stealing would be helpful here. If you look at the resulting PNGs in my project, I think it can be said that optimally you should be able to have the number of concurrent tests equal to maxParallelForks as much as possible to take advantage of the parallelism made available from your server. 

- As it was said earlier, binning the long running tests would be a good idea. This is not shown in my sample screenshots, but for my build at work I have observed several long running tests executing back to back. This is sort of a critical path, hindering speedup.

- A maintenance free, dynamic way to determine these bins might be to read something like the previous nighly junit test report. A test task may be assigned the longest N tests, where N equals the maxParallelForks value. I havent implemented this yet, but I intend to.

**Comment by oehme** [comment]
Sorting by execution time actually sounds like a cheap idea to improve this on the Gradle side. Then you only pay the cost of inefficient scheduling once.

**Comment by 4n00p** [comment]
What do you think about this idea? Any guidance?  I think it would offer better speedup than moving the longest running tests to a single Test task:

The optimal case is one that shows the ideal speedup (adding a 2nd process reduces duration by 50%).  This is when each test is the same duration (as seen in [screenshots/Figure_3.png](https://raw.githubusercontent.com/4n00p/gradle_log_test_timeline/master/screenshots/Figure_3.png) )  

I want to get my build as close to the optimal case above.  Using an older JUnit test report, start by constructing a list with test class name + duration and sort by the duration.  Then, dynamically create Test tasks that each only include up to N tests, where N == maxParallelForks and the test class exists in the sourceSet (the JUnit test report and list won't have subproject information).  I will have to make sure that every task has its N "slots" full with tests.  If many tests are similar in  duration, then I can have single Test task contain 2*N, etc.  

This would create many Test tasks within a subproject, but hopefully the overhead from that would be minimal.
Why so many Test tasks? Simply to prevent queuing of multiple, long-running tests (the original problem in this ticket :)  )  I'm sure you all know this, but here is my example:  

```
If maxParallelForks = 2, and the sourceSet contains JUnit tests:
 Test1,Test2,Test3,Test4.
Assume these tests happen to be in sorted order of duration.
If there is one task, Task1, with sourceSet:
Test1 = duration of 100 seconds
Test2 =                     90 seconds
Test3 =                     80 seconds
Test4 =                     70 seconds

It may result in a total task duration of 190 seconds, depending on 
how the JUnit tests are assigned:
order of execution  ---->
0 seconds                  100 seconds               190 seconds
         Task1:Test1 (100 sec) | Task1:Test2 (90 sec)
         Task1:Test3 (80 sec)   | Task1:Test4 (70 sec)

If maxParallelForks = 2 and there are 2 Tasks:

Task 1:
Test1 = duration of 100 seconds
Test2 =                    90 seconds

Task 2:
Test3 =             80 seconds
Test4 =             70 seconds

order of execution  ---->
0 seconds                  100 seconds                      180 seconds
         Task1:Test1 (100 sec) | Task2:Test2 (80 sec)
         Task1:Test2 (90 sec)   | Task2:Test4 (70 sec)

Task1 and Task2 complete these Tests in 180 seconds 
(assuming not much overhead switching from Task1 to Task2)
```
The tests are assigned from the sourceSet in what seems to be a round-robin assignment to workers from the source tree...? Well, no matter, it is a set and no order is guaranteed.  
In summary:
- create a list of JUnit test names, sorted by duration 
- per subproject, dynamically create Test tasks with "include" statements. The include should take the first N tests (that are in the sourceSet) then the next Test task takes the next N tests,etc.

Another approach I was thinking about was to use JavaExec to "blast" the Junit tests in parallel...then somehow collect the html output into a single report.

**Comment by oehme** [comment]
I didn't mean to hack this into your own build, but actually change Gradle to do this. Currently the tests are executed in the order they are visited on the file system. The previous test report could be used to reorder them based on execution time. We already do a similar reordering for [rerunning failed tests first](https://github.com/gradle/gradle/blob/master/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/RunPreviousFailedFirstTestClassProcessor.java). A contribution for that would be very welcome.

**Comment by dellgreen** [comment]
sounds great, but it would be nice to have a solution that benefits everyone including Desktop users and CI servers. Either way it all sounds promising

**Comment by 4n00p** [comment]
That seems like a good idea, since you won't have the extra overhead of starting many different tasks  (that execute no more than the maxParallelForks value). There doesn't seem to be much overhead, though.  Probably the biggest gain is less clutter in your build script.  
I ended up with a quick demo solution here:
https://github.com/4n00p/gradle_improve_concurrency  

The TestReport task, 'run_test_N', calls, in order:

> :Subproject_One:test_a 
> :Subproject_One:test_aa 
> :Subproject_One:test_aaa 
> 
> :Subproject_Two:test_a 
> :Subproject_Two:test_aa 
> :Subproject_Two:test_aaa 

Note that I never bothered to parse the JUnit test, sort by duration and group by subproject.  I just "mocked" the sorted list.  I wanted a thread through with the gradle portion first.  
Any thoughts on this "prototype" project would be appreciated!  

So how much work would there be to add reordering to the gradle build as you suggest?

**Comment by oehme** [comment]
It's not much, see the class I linked earlier. Getting to know the code base and writing some good tests would probably take more time than writing the actual ordering logic.

**Comment by 4n00p** [comment]
Ok. I started working on the contrib. Do you folks have gitter chat room for contributors?

**Comment by oehme** [comment]
@eriwen can invite you to our community slack.

**Comment by 4n00p** [comment]
@oehme @eriwen Could somebody add me to slack? I haven’t received any notification yet. Sorry to be pushy. Thanks.

**Comment by eriwen** [comment]
@4n00p Please use [this link](https://join.slack.com/t/gradle-community/shared_invite/enQtNDE3MzAwNjkxMzY0LTYwMTk0MWUwN2FiMzIzOWM3MzBjYjMxNWYzMDE1NGIwOTJkMTQ2NDEzOGM2OWIzNmU1ZTk5MjVhYjFhMTI3MmE) to join community slack.

**Comment by aleksandarsusnjar** [comment]
Placing tests into multiple queues and redistributing them as situations change is both complex and unneeded. The ideal case is to simply have a shared queue (or other kind of pool) of work items, from which the workers can get them and execute them as soon as they become free.

This would affect the maxParallelForks though, if it is to be kept, as mechanism as described does not naturally support such limits. I personally don't need maxParallelForks restrictions but those who do probably need it to throttle load on other systems or similar. To support it a mechanism as I described would have to be able to NOT get a test/work from the "queue" if it were to violate this limit. As such, it cannot really be a queue any more, but another kind of pool.

**Comment by tcfurrer** [comment]
It is completely legal and sane for a test suite to contain a large number of tests with runtimes varying from milliseconds to minutes.  Not every project organizes its test suites into runtime-based buckets, and for good reasons.  Sometimes there are other dimensions of categorization that are more important, and it is undesirable to complicate test suite definitions by creating a cross product of runtime categorization with some other categorization.

Also, there are many reasons why using prior runtime data isn't always great for splitting up work lists.  Test runtimes can vary for various reasons (cpu load, uncontrained randomized testing, external dependencies, etc).  Also the list of tests in a suite can change between runs.  For example on a large and highly active project, one may quite frequently be doing a repoistory update which changes/adds/removes a fair number of tests.  And so on.

It seems to me that work stealing is a good choice for Gradle, and will handle a wide variety of usage cases better than most other approaches.  I'm also not convinced that it's too complex or difficult to do.  Yes it's not trivial either, but multi-threaded performance is one thing which really makes Gradle stand out from alternative tools, and so it seems like a good place to make long term investments.

**Comment by jacattau** [comment]
I just started trying to get Geb and Spock tests running in parallel and I'm running into this frustrating issue.  I'll have 6 test classes and 4 forks.  The first 4 forks run in parallel just fine, but when the first 2 classes finish before the last 2, it still waits until the last 2 are done to spin up other workers to complete the tests.  

This discussion sounds exactly like what I'm looking for.  Where are we at on the progress?

**Comment by samzurcher** [comment]
@4n00p you mentioned in September of last year, that you have started working on this. Any progress so far? My team would be very interested in that feature :-).

**Comment by samzurcher** [comment]
@oehme @eriwen are you guys working on this issue? I think many people would be interested in that feature.

**Comment by oehme** [comment]
We are not working on this at the moment. I know that better test concurrency would be great, but it's not at the top of our priority list at the moment.

**Comment by samzurcher** [comment]
@oehme Would you guys accept a pull request that tries to tackle the problem by using historic execution time data and based on that organize the tests into _balanced_ sets? Would you by any chance know where the relevant entry point is in the code?

**Comment by oehme** [comment]
Sure, you can have a look at [RunPreviousFailedFirstTestClassProcessor](https://github.com/gradle/gradle/blob/master/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/processors/RunPreviousFailedFirstTestClassProcessor.java) for inspiration. You don't need to split it into sets, just sort by most expensive.

**Comment by dagguh** [comment]
I second @tcfurrer 's [comment](#issuecomment-446434572).
Historical test durations only add complexity and introduce new problems. How to share the history across CI and local dev? What if it's too random?

It It could be a single producer, multiple consumers queue. The shortest possible wall clock duration would emerge naturally.

PS. Do we have to fork entire JVMs? `Executors.newFixedThreadPool` should be more than enough for most use cases. Forking JVMs adds overhead (e.g. startup, less JIT optimizations) and makes resource sharing less efficient (e.g. in-memory caches).

**Comment by oehme** [comment]
Forking VMs is necessary for many reasons, like isolating the build from badly behaved tests (which could make the daemon unstable), being able to cancel tests (test frameworks generally swallow interrrupts), controlling resources etc.

I guess by queue you mean doing one class at a time? That would be inefficient, since most test frameworks (and test code bases) have some pretty substantial setup cost for each test run. Chunking tests into buckets is pretty much unavoidable if you want good performance.

Having a number of buckets greater than the number of processors, but smaller than the number of classes and then consuming them as processors become available could be a good middle ground between optimal scheduling and overhead.

**Comment by dagguh** [comment]
Ok, I see the point of forks 👍 

> I guess by queue you mean doing one class at a time?

A queue is just a structure for tasks (e.g test classes to run or test methods to run). Spawn N consumer threads/forks and each takes one task at a time, completes it and asks the queue for another one, until the queue is empty. No explicit buckets necessary.

Also note that many kinds of tests are not bottleneck by CPU, but by disk or network, so we should allow for higher concurrency without limiting the user to the CPU core count.

**Comment by aleksandarsusnjar** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}                                                                                                                                        This can still work automatically, with queues. The consumers would just get batches of tests to execute (i.e. not individual tests) and can wait (for just a little while) for more tests to become available if not enough are there.This "batching" can ideally occur on the consuming end because then there is most control available but it is also possible to go through a "batching producer". Otherwise there is no reason to limit this to a single producer. Also, it does not have to be an ordered queue at all, just an unordered pool of tests to be run.                                                                                                                                                                          From: notifications@github.comSent: July 5, 2019 11:28 a.m.To: gradle@noreply.github.comReply to: reply@reply.github.comCc: github@susnjar.net; comment@noreply.github.comSubject: Re: [gradle/gradle] Improve parallel tests concurrency (#2669)  Ok, I see the point of forks 👍



I guess by queue you mean doing one class at a time?



A queue is just a structure for tasks (e.g test classes to run or test methods to run). Spawn N consumer threads/forks and each takes one task at a time, completes it and asks the queue for another one, until the queue is empty. No explicit buckets necessary.

Also note that many kinds of tests are not bottleneck by CPU, but by disk or network, so we should allow for higher concurrency without limiting the user to the CPU core count.

**Comment by oehme** [comment]
> Spawn N consumer threads/forks and each takes one task at a time

That doesn't work for the reason mentioned above - Kicking off a whole JUnit run just to run a single test task is highly inefficient.

> The consumers would just get batches of tests to execute

That's what I pointed out above. The problem is choosing the right batch size, because as mentioned earlier, a JUnit run can have significant overhead (e.g. if you have RunListeners that spin up services).

**Comment by tcfurrer** [comment]
On the one end, there could be a large number of test classes which each take 10ms to run.  On the other end, there could be a single testcase that takes an hour to run.  And since runtime isn't known beforehand (as discussed already), there is basically no right heuristic batch size to choose, which will work well for any test suite thrown at it.

If you choose one test at a time from a shared queue, the overhead of process communication starts to dominate when there are lots of fast tests.

If you choose even just two tests at a time, you might inadvertently create a long pole worker (i.e. a worker that continues running for significant wall clock time after all other workers are complete) if long running tests are being chosen.

Assuming Junit 5 is intended for general purpose use across a wide variety of test suites, I don't think it would be safe to assume that the aforementioned cases are only fictitious examples.  I believe it is realistic for Junit 5 to encounter test suites across the spectrum of test runtimes, total suite runtimes, and variation in test run times.

So I think what is needed is a solution that can deal with a wide variety of suites while also ensuring that a test suite's total parallel runtime is not sub-optimal because one worker continues on for a long time while other workers do nothing, even though there are multiple test classes remaining to be run.

**Comment by aleksandarsusnjar** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}  Well, one possibility is to  use dummy/delegate tests that consume a real test from the pool when they get to run. If there is nothing more to run for a (very little while) they can exit. One issue with this is test statistics, though. They could add a little bit more time to the end but not much.It would be ideal if there is a way to NOT fix a number of tests per run/JVM (with respect to this, this may need to be done otherwise), but simply feed test as they are found - iterate through them.                                                                                                                                                                          From: notifications@github.comSent: July 5, 2019 8:33 p.m.To: gradle@noreply.github.comReply to: reply@reply.github.comCc: github@susnjar.net; comment@noreply.github.comSubject: Re: [gradle/gradle] Improve parallel tests concurrency (#2669)  On the one end, there could be a large number of test classes which each take 10ms to run.  On the other end, there could be a single testcase that takes an hour to run.  And since runtime isn't known beforehand (as discussed already), there is basically no right heuristic batch size to choose, which will work well for any test suite thrown at it.
If you choose one test at a time from a shared queue, the overhead of process communication starts to dominate when there are lots of fast tests.
If you choose even just two tests at a time, you might inadvertently create a long pole worker (i.e. a worker that continues running for significant wall clock time after all other workers are complete) if long running tests are being chosen.
Assuming Junit 5 is intended for general purpose use across a wide variety of test suites, I don't think it would be safe to assume that the aforementioned cases are only fictitious examples.  I believe it is realistic for Junit 5 to encounter test suites across the spectrum of test runtimes, total suite runtimes, and variation in test run times.
So I think what is needed is a solution that can deal with a wide variety of suites while also ensuring that a test suite's total parallel runtime is not sub-optimal because one worker continues on for a long time while other workers do nothing, even though there are multiple test classes remaining to be run.

**Comment by dagguh** [comment]
Personally I suffer from the long-pole problem you described. Adding one 40 seconds long test changed the total test time from 12 minutes (highly concurrent) to 1 hour (barely concurrent).

I also believe that the solution should be flexible enough to accomodate both classes of use cases, even if it means the user has to configure more.

**Comment by markrmiller** [comment]
You might look at how https://labs.carrotsearch.com/randomizedtesting.html handles this for ant in https://github.com/randomizedtesting/randomizedtesting/tree/master/junit4-ant

It keeps a history of the the execution time for the past n test runs for each test and allows you to use that info to balance test distribution across jvms, allowing you to schedule longer running tests first and balance runtimes across jvms. You can commit a seed stats file with point in time stats to the repo for the first run where a build has no stats and update that infrequently, essentially a best effort hint file for a fresh env or when jenkins doesn't repeat envs necessarily.

This will statically assign tests across jvms, but also keep a percentage of tests to assign dynamically as jvms free up.

It's super dope in comparison to no ordering round robin.

**Comment by roded** [comment]
We're affected by the long poling issue as well when running our unit and integration tests via Gradle. The price payed for unit tests is in the order of minutes while for our integration tests it's more like an hour or so of excess clock time which could be parallelized. Is this issue stuck due to design considerations? Is a contribution required?

**Comment by hansi-b** [comment]
We are also affected by this, since our mix of tests sports runtimes from minutes down to next to nothing.

It seems to me that merely sorting by most expensive would result in a necessarily suboptimal result when used with the current round robin approach of the `MaxNParallelTestClassProcessor`: Intuitively speaking, for each job assigned to the first processor, the second (or nth) processor gets a smaller one. This would introduce a systematic imbalance which probably increases the more skewed the runtime distribution for the test classes becomes.

I have pushed to my fork a first sketch in which I use the recorded test class runtimes as loads in the processor to distribute the tests. There are no tests yet (the existing ones run through), and I haven't tried it out on any actual suite.

I'd be grateful for comments with regard to whether this approach might be worthwhile to follow up on as a pull request.

**Comment by hansi-b** [comment]
For the moment, I have pedalled back from the previous idea of keeping track of processor load. [My pull request for this issue](https://github.com/gradle/gradle/pull/12012) uses the simple duration-sorting approach [proposed by @oehme](https://github.com/gradle/gradle/issues/2669#issuecomment-401061758), together with a minor modification of the round robin allocation.

This is a very small change:
* It maintains the invariant that test class numbers across processors will not differ by more than one. 
* It does not change the current behaviour when no previous test durations are available.

Some brief simulations (https://github.com/hansi-b/Long-Poles/blob/master/LongPoleTests.ipynb) suggest this simple heuristic improves the overall time quite dramatically. Obviously, this is still a static pre-allocation and assumes that test runtime information is available and reliable.

**Comment by hansi-b** [comment]
Wondering if someone more familiar with the process has any comment: The pull request pertaining to this issue is basically done (AFAICS). Last review-related code changes occurred in January. Apart from waiting (and rebasing on merge conflicts), there anything else I should or can do?

**Comment by genman** [comment]
Looking at @hansi-b 's patch, I think the biggest issue is that our build server would have to maintain some sort of history of executing all the tests before it knew how to schedule everything, and that doesn't seem feasible when building from scratch different forks and versions. I suppose it's fine in some cases and probably less impact, but I would say the design needs fixing.

The list of pending tests should be in a queue, and they should get executed using a worker JVM as soon as that JVM is idle. We see something like this, where there are 8 works but only one executing anything.
```
<============-> 97% EXECUTING [16m 9s]
> foo:test > 181 tests completed, 50 failed, 36 skipped
> IDLE
> IDLE
> IDLE
> IDLE
> IDLE
> IDLE
> :foo:test > Executing test BarIntegrationTest
> IDLE
```

**Comment by dagguh** [comment]
The most efficient model is work stealing. Workers could take tests from a shared pool whenever they're free/idle. It's a **pull** model. The problem is: Gradle **pushes** tests into workers. Perhaps it was the easiest way to integrate JUnit years ago. Gotta switch to a pull model.

**Comment by hansi-b** [comment]
Sorry for the commit notification muck hill here; I have removed the resolvable issue reference from the commits, and I will probably no longer have the time to update the pull request anyway.
We are currently running our tests on a 6.6.1 build using this fix, and that works fine for our use case (cuts time from around 1h 20m to 50m), but a more general solution like work stealing would clearly be preferable.

**Comment by spyro2000** [comment]
For us, i most cases the parallel execution is much slower than using a single run. Even on a 8 core machine. Problem is that the workers are stale after "their" test has been executed instead of immediately running the next one.

**Comment by jbduncan** [comment]
This issue sounds rather valuable, regardless of whether it's implemented using work-stealing or by sorting tests by previous run times.

**Comment by markrmiller** [comment]
Yeah, you have to hold back tests rather than feeding them all up front. You also want to run the longest tests first. I the gradle test method into another test target once. In that case, the slow tests also didn't actually use many resources, so I designated most parallel JVMs to get large tests, but a couple would get fast tests right away as well. When the large were done, all JVMs would take slow tests. I put a queue in front of each JVM and gave the tests weights depending on length and ensured the queue was fed below a certain wait to ensure some tests were always queued up to go next. The first run, you won't have test times (I used annotations) but after that you can keep a file with avg times per test. So two threads sat on the tests, fast and slow tests, and dribbled them off to be filtered into a JVM queue until that queues min weight was hit, and then would block until the weight went under the min weight before unblocking. That gets you close to optimal times, maxing resources to a decent extent and ensuring JVMs don't end up testless at the end with a pile of the remaining tests lined up on a couple or one JVM.

**Comment by axelfontaine** [comment]
Is there any progress on this? 

- This has now been open for almost 5 years (!).
- It's obviously impacting a very large number of users.
- The solution seems pretty clear to everyone involved.

Dear Gradle team, what else do you need to act on this?

**Comment by zeners** [comment]
I was able to replace the round robin algorithm with own, based on statistics from previous runs, but fot this, i have to patch 3 Files and one class to override.

I saved more than 50% time in summary.

with the statistics from previuos runs, it was easy: 
- sorting known Tests by time descening, so most time consuming testclass is first assigned (patching RunPreviousFailedFirstTestClassProcessor)
- use processor with least time planned from assigned test-classes (patching MaxNParallelTestClassProcessor)

the used algorithm fall back to round robin, if no statistics are provided 

i would like to see waiting for idle processor, so real used time will come to hand, but for good results, you need sorted classes by estimted consume time

if you like my approch, i try to make a PR of my Changes:

- org.gradle.api.tasks.testing.Test:
  to accept a map of known times per class
- org.gradle.api.internal.tasks.testing.detection.DefaultTestExecuter:
  deploy the known Times to the following classes:
  - org.gradle.api.internal.tasks.testing.processors.RunPreviousFailedFirstTestClassProcessor: 
    sort otherTestClasses by known Times respecting testFilter
  - org.gradle.api.internal.tasks.testing.processors.MaxNParallelTestClassProcessor:
    remember assigned times and use processor with lowest ones

**Comment by hansi-b** [comment]
@zeners, interesting! Sounds similar to the PR I had prepared back in January 2020: https://github.com/gradle/gradle/pull/12012, if that is of any interest for comparison. The PR went through some reviews, but got no further attention from the Gradle team - at least as far as I could see - and eventually went stale.

**Comment by dagguh** [comment]
Tracking time is unnecessarily complex and flaky. Just use work stealing.

**Comment by dellgreen** [comment]
As discussed before, any approach that depends on previous runs is really the wrong approach, for all the CI systems or developers that perform clean builds every time. Work stealing theoretically is a better long term solution as discussed in the past.

**Comment by zeners** [comment]
thanks for discussing it

Work stealing is not enough: the tests should be sorted first by estimated run time. Otherwise the solution will miss the optimal runtime.

I like to split the cause in the two problems of sorting test-tasks (currently fifo by LinkedHashSet) and assigning to processes (currently round-robin without work-stealing).

Currently it's hard to replace the 2 hard-coded aspects (as you have to touch/copy 4 classes).

All i wish is the ability to easy replace these two aspects with custom ones. If gradle offer work-stealing, i like to prefer it. But i need the ability to sort the tests based by my own statistics (in my case: it came from the average of the last 50 runs, provided by the CI-system, already done, no need to be part of gradle to supply it). 

I can replace the whole DefaultTestExecutor (as i currently do), but i prefer to only replace the to aspects, leaving the rest untouched.

Both can be accomplished by accepting an alternate processor in DefaultTestExecutor via a Factory<TestClassProcessor>.

@dellgreen i try to provide (simulated) processor work timings, to show the impact of presorted test-classes. Nevertheless i like to see work-steeling implemented. I see presorting as a way for optional optimizing beyond processor-assigning.

**Comment by dellgreen** [comment]
@zeners I don't doubt that this is the right solution for you, but given how we have been discussing this ticket for nearly 6 years now with statical solutions being proposed before and gained little traction. Any chance of success in a change in Gradle I feel needs to benefit as many as possible with zero effort/infrastructure to users to benefit from the first time they run.  I personally feel the 'most' optimal solution is less important at this stage, with preference to a solution that on 'average' produces a better result for all users. Hence, I feel efforts should first be targeted at work stealing to stand any chance of being merged.  Please don't take this in any way to hinder your efforts, I just feel we will be discussing this for another 6 years unless a simpler solution is proposed

**Comment by zeners** [comment]
simulated data:
current (without sorting & assigning):
PT5M47.386000002S
PT7M14.929000001S
PT7M18.972000003S
PT8M27.132000003S
PT8M31.229000004S
PT8M34.101000001S
PT8M54.118000002S
PT10M43.693000002S
PT11M34.69S
PT12M38.649000001S
PT13M27.359000002S
PT22M27.321000003S

only assigning (same as work-steeling under ideal circumstances)
PT9M10.856S
PT9M13.709000004S
PT9M26.650000001S
PT9M39.776S
PT9M40.255000009S
PT10M3.089S
PT10M19.738S
PT10M20.828000001S
PT11M10.716S
PT11M13.839000005S
PT11M56.270000003S
PT13M23.853000001S

sorted and assigning:
PT10M27.851000024S
PT10M27.927S
PT10M27.956S
PT10M27.97S
PT10M28.005S
PT10M28.055S
PT10M28.139S
PT10M28.144S
PT10M28.145S
PT10M28.151S
PT10M28.865S
PT10M30.371S

my conclusion for my scenario:
- current gradle work:
  100 % time 22:27
- without sorting, but gradle can handle standalone:
  60 % time 13:24
- including sorting:
  47 % time 10:30

i'm able to make a PR that in my scenario 40% time safe for everyone and up to 53% time safe, if i can supply time statics

@dellgreen: do you like to see this as a PR?
the solution IS simple and i imagine a ci-plugin, providing the statistics (it already has)

**Comment by zeners** [comment]
currently testing the following stealing concept:
- Stealer class, using newSingleThreadExecutor to add, remove and trySteal (find first not running Stealable)
- Stealable class implements and wrapping TestClassRunInfo with volatile running flag, synchronized switching state and stealer
- MaxNParallelTestClassProcessor wrapping testClass with Stealable and adding it to Stealer before processor.processTestClass
- TestWorker 
  - processTestClass: test (testClass not beeing Stealable || try switch running flag including async remove from Stealer) before forwarded
  - stop: trysteal, if successful run it and submit stop again, otherwise do stop

so it should have minimum overhead, except stealing, that is serialized via singleThreadExecutor 
- main runnings only synchonized per testClass (so no locking between test-classes, if startet in parallel)
- start running testClasses removed async from Stealer, so no locking 
- mostly find first Stealable on first element in list

hope this sounds like a working plan

**Comment by zeners** [comment]
testing shows up: gradle and junit5 is funny: JUnitPlatformTestClassProcessor uses CollectAllTestClassesExecutor. 
So stealing have to be delegated to this processor (it seem's he didn't like that, but i think, i can handle that)

to make it right, it needs a stealing handling strategy controlled by the Testframework, deciding between 
 - self-handling (for junit5), 
 - handled by TestWorker (junit4) or 
 - be complete incompatible (other not testet/checked) and 
 - the option to be disabled by user

i don't like to summarize different aspects into one commit/PR, i would split it up into multiple PRs. more, if i have something to present

thanks for input

PS: sry for self commenting

**Comment by bertvanbrakel** [comment]
I vote for the ability to maintain a timing file and based off that distribute work. 

Don't let the perfect get in the way of better. This request was open in 2017, so clearly perfect got in the way.

The work stealing ticket is still under discussion, works only for junit4, so I guess another few years and we might have it.

I think being able to provide ones own timing file would have been a quicker solution. Just need to split it up into the number of processes and away you go. Up to the user to scrap/generate the file per env,

We're currently got hour long plus builds, we should be able to get them down to less than half, as for most of the time the processes are sitting idle.

**Comment by zeners** [comment]
> The work stealing ticket is still under discussion, works only for junit4, so I guess another few years and we might have it.

work-stealing with junit5 works always with another commit ([zeners:gradle#JUnit5_TestClass_Stealing](https://github.com/zeners/gradle/tree/JUnit5_TestClass_Stealing)) and sorting based on previous runs is also available as commit ([zeners:gradle#TestSortingAndAssigning](https://github.com/zeners/gradle/tree/TestSortingAndAssigning)). I'm happy about comments for they. Only junit-ng should be discussed, but the solution for junti5 is simple to adapt.

I prefer to evolve the solution, rather to make an all in one solution commit. Both aspects (work-stealing and sorting) are boosts for the overall-runtime. But not all scenarios are interested in optimized overall-runtime and that should be respected too.

> I think being able to provide ones own timing file would have been a quicker solution. Just need to split it up into the number of processes and away you go. Up to the user to scrap/generate the file per env,

that's exactly the way my TestSortingAndAssigning-branch handle it and it's up to you to combine it with work-stealing or not.

**Comment by felixscheinost** [comment]
Seems Gradle Enterprise has [various features](https://gradle.com/gradle-enterprise-solutions/test-distribution/) to make test execution faster. 

That explains why it isn't in Gradle's interest to fix this issue. Sad.

**Comment by dellgreen** [comment]
> Seems Gradle Enterprise has [various features](https://gradle.com/gradle-enterprise-solutions/test-distribution/) to make test execution faster.
> 
> That explains why it isn't in Gradle's interest to fix this issue. Sad.

Looks like that is centered around historical execution speed sorting, so could be some scope for an open source work stealing addition/plugin test runner maybe?

**Comment by zeners** [comment]
> could be some scope for an open source work stealing addition/plugin test runner maybe
this was one of my intention: make an entrypoint for plugins: [TestClassSortAndAssignStrategy](https://github.com/zeners/gradle/blob/TestSortingAndAssigning/subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/TestClassSortAndAssignStrategy.java)
